File: examples/c++/simpleExample.cpp
/* ----------------------------------------------------------------------------

 * QuadricSLAM Copyright 2020, ARC Centre of Excellence for Robotic Vision, Queensland University of Technology (QUT)
 * Brisbane, QLD 4000
 * All Rights Reserved
 * Authors: Lachlan Nicholson, et al. (see THANKS for the full author list)
 * See LICENSE for the license information

 * -------------------------------------------------------------------------- */

/**
 * @file System.cpp
 * @date Apr 14, 2020
 * @author Lachlan Nicholson
 * @brief code to operate the quadric slam library on a dataset
 */

/**
 * A simple example with multiple camera positions and quadric landmarks
 *  - graph provided with true box measurements, from projecting quadric into image at each frame
 *  - graph provided with true odometry measurements
 *  - initial estimate given perturbed trajectory
 *  - initial estimate given true quadrics
 */

#include <gtsam_quadrics/geometry/ConstrainedDualQuadric.h>
#include <gtsam_quadrics/geometry/AlignedBox2.h>
#include <gtsam_quadrics/geometry/QuadricCamera.h>
#include <gtsam_quadrics/geometry/BoundingBoxFactor.h>

#include <gtsam/nonlinear/NonlinearFactorGraph.h>
#include <gtsam/nonlinear/Values.h>
#include <gtsam/nonlinear/LevenbergMarquardtOptimizer.h>
#include <gtsam/nonlinear/LevenbergMarquardtParams.h>
#include <gtsam/geometry/Pose3.h>
#include <gtsam/geometry/CalibratedCamera.h>
#include <gtsam/slam/BetweenFactor.h>
#include <gtsam/slam/PriorFactor.h>

#include <random>
#include <string>
#include <iostream>

#define ODOM_SD 0.01
#define BOX_SD 3

using namespace gtsam;
using namespace std;

int main(void) {

    // create empty graph / estimate
    NonlinearFactorGraph graph;
    Values initialEstimate;

    // define calibration
    boost::shared_ptr<Cal3_S2> calibration(new Cal3_S2(525.0, 525.0, 0.0, 320.0, 240.0));

    // define noise models 
    boost::shared_ptr<noiseModel::Diagonal> priorNoiseModel = noiseModel::Diagonal::Sigmas(Vector6::Ones()*1e-1);
    boost::shared_ptr<noiseModel::Diagonal> odomNoiseModel = noiseModel::Diagonal::Sigmas(Vector6::Ones()*ODOM_SD);
    boost::shared_ptr<noiseModel::Diagonal> boxNoiseModel = noiseModel::Diagonal::Sigmas(Vector4::Ones()*BOX_SD);

    // define key poses
    vector<Pose3> trajectory;
    trajectory.push_back(gtsam::CalibratedCamera::LookatPose(Point3(10,0,0), Point3(), Point3(0,0,1)));
    trajectory.push_back(gtsam::CalibratedCamera::LookatPose(Point3(0,-10,0), Point3(), Point3(0,0,1)));
    trajectory.push_back(gtsam::CalibratedCamera::LookatPose(Point3(-10,0,0), Point3(), Point3(0,0,1)));
    trajectory.push_back(gtsam::CalibratedCamera::LookatPose(Point3(0,10,0), Point3(), Point3(0,0,1)));
    trajectory.push_back(gtsam::CalibratedCamera::LookatPose(Point3(10,0,0), Point3(), Point3(0,0,1)));

    // define quadrics
    vector<ConstrainedDualQuadric> quadrics;
    quadrics.push_back(ConstrainedDualQuadric(Pose3(), Vector3(1.,2.,3.)));
    quadrics.push_back(ConstrainedDualQuadric(Pose3(Rot3(), Point3(0.1,0.1,0.1)), Vector3(1.,2.,3.)));

    // add prior on first pose
    PriorFactor<Pose3> priorFactor(Symbol('x',0), trajectory[0], priorNoiseModel);
    graph.add(priorFactor);

    // add trajectory estimate
    for (unsigned i = 0; i < trajectory.size(); i++) {

        // add a perturbation to initial pose estimates to simulate noise
        Pose3 perturbedPose = trajectory[i].compose(Pose3(Rot3::rodriguez(0.1,0.1,0.1), Point3(0.1,0.2,0.3)));
        initialEstimate.insert(Symbol('x', i), perturbedPose);
    }

    // add quadric estimate
    for (unsigned i = 0; i < quadrics.size(); i++) {
        initialEstimate.insert(Symbol('q', i), quadrics[i]);
    }
    
    // add relative poses to graph as odometry
    for (unsigned i = 0; i < trajectory.size()-1; i++) {
        Pose3 relativePose = trajectory[i].between(trajectory[i+1]);
        BetweenFactor<Pose3> bf(Symbol('x', i), Symbol('x', i+1), relativePose, odomNoiseModel);
        graph.add(bf);
    }

    // reproject true quadrics into each true pose
    for (unsigned j = 0; j < quadrics.size(); j++) {
        for (unsigned i = 0; i < trajectory.size(); i++) {
            DualConic conic = QuadricCamera::project(quadrics[j], trajectory[i], calibration);
            AlignedBox2 bounds = conic.bounds();
            BoundingBoxFactor bbf(bounds, calibration, Symbol('x', i), Symbol('q', j), boxNoiseModel);
            graph.add(bbf);
        }
    }

    // define lm parameters
    LevenbergMarquardtParams parameters;
    parameters.setVerbosityLM("SUMMARY"); // SILENT = 0, SUMMARY, TERMINATION, LAMBDA, TRYLAMBDA, TRYCONFIG, DAMPED, TRYDELTA
    parameters.setMaxIterations(100); 

    // build optimiser
    LevenbergMarquardtOptimizer optimizer(graph, initialEstimate, parameters);

    // optimise the graph
    Values result = optimizer.optimize();

    return 1;
}

File: examples/python/simple_example.py
"""
QuadricSLAM Copyright 2020, ARC Centre of Excellence for Robotic Vision, Queensland University of Technology (QUT)
Brisbane, QLD 4000
All Rights Reserved

See LICENSE for the license information

Description: example problem using python interface 
Author: Lachlan Nicholson (Python)
"""

# import standard libraries
import os
import sys
import numpy as np
import matplotlib.pyplot as plt

# import gtsam and extension
import gtsam
import gtsam_quadrics

if __name__ == '__main__':

    # declare noise estimates
    ODOM_SIGMA = 0.01
    BOX_SIGMA = 3

    # define shortcuts for symbols
    X = lambda i: int(gtsam.symbol(ord('x'), i))
    Q = lambda i: int(gtsam.symbol(ord('q'), i))

    # create empty graph / estimate
    graph = gtsam.NonlinearFactorGraph()
    initial_estimate = gtsam.Values()

    # define calibration 
    calibration = gtsam.Cal3_S2(525.0, 525.0, 0.0, 160.0, 120.0)

    # define noise models
    prior_noise = gtsam.noiseModel_Diagonal.Sigmas(np.array([1e-1]*6, dtype=np.float))
    odometry_noise = gtsam.noiseModel_Diagonal.Sigmas(np.array([ODOM_SIGMA]*3 + [ODOM_SIGMA]*3, dtype=np.float))
    bbox_noise = gtsam.noiseModel_Diagonal.Sigmas(np.array([BOX_SIGMA]*4, dtype=np.float))

    # define poses 
    poses = []
    poses.append(gtsam.SimpleCamera.Lookat(gtsam.Point3(10,0,0), gtsam.Point3(), gtsam.Point3(0,0,1)).pose())
    poses.append(gtsam.SimpleCamera.Lookat(gtsam.Point3(0,-10,0), gtsam.Point3(), gtsam.Point3(0,0,1)).pose())
    poses.append(gtsam.SimpleCamera.Lookat(gtsam.Point3(-10,0,0), gtsam.Point3(), gtsam.Point3(0,0,1)).pose())
    poses.append(gtsam.SimpleCamera.Lookat(gtsam.Point3(0,10,0), gtsam.Point3(), gtsam.Point3(0,0,1)).pose())
    poses.append(gtsam.SimpleCamera.Lookat(gtsam.Point3(10,0,0), gtsam.Point3(), gtsam.Point3(0,0,1)).pose())

    # define quadrics
    quadrics = []
    quadrics.append(gtsam_quadrics.ConstrainedDualQuadric(gtsam.Pose3(), np.array([1.,2.,3.])))
    quadrics.append(gtsam_quadrics.ConstrainedDualQuadric(gtsam.Pose3(gtsam.Rot3(), gtsam.Point3(0.1,0.1,0.1)), np.array([1.,2.,3.])))

    # add prior on first pose
    prior_factor = gtsam.PriorFactorPose3(X(0), poses[0], prior_noise)
    graph.add(prior_factor)

    # add trajectory estimate
    for i, pose in enumerate(poses):

        # add a perturbation to initial pose estimates to simulate noise
        perturbed_pose = poses[i].compose(gtsam.Pose3(gtsam.Rot3.Rodrigues(0.1,0.1,0.1), gtsam.Point3(0.1,0.2,0.3)))
        initial_estimate.insert(X(i), perturbed_pose)

    # add quadric estimate
    for i, quadric in enumerate(quadrics):
        quadric.addToValues(initial_estimate, Q(i))

    # add relative poses to graph as odometry
    for i in range(len(poses)-1):
        relative_pose = poses[i].between(poses[i+1])
        odometry_factor = gtsam.BetweenFactorPose3(X(i), X(i+1), relative_pose, odometry_noise)
        graph.add(odometry_factor)

    # reproject true quadrics into each true pose
    for j, quadric in enumerate(quadrics):
        for i, pose in enumerate(poses):
            conic = gtsam_quadrics.QuadricCamera.project(quadric, pose, calibration)
            bounds = conic.bounds()
            bbf = gtsam_quadrics.BoundingBoxFactor(bounds, calibration, X(i), Q(j), bbox_noise)
            graph.add(bbf)

    # define lm parameters
    parameters = gtsam.LevenbergMarquardtParams()
    parameters.setVerbosityLM("SUMMARY") # SILENT = 0, SUMMARY, TERMINATION, LAMBDA, TRYLAMBDA, TRYCONFIG, DAMPED, TRYDELTA : VALUES, ERROR 
    parameters.setMaxIterations(100)
    parameters.setlambdaInitial(1e-5)
    parameters.setlambdaUpperBound(1e10)
    parameters.setlambdaLowerBound(1e-8)
    parameters.setRelativeErrorTol(1e-5)
    parameters.setAbsoluteErrorTol(1e-5)

    # create optimizer
    optimizer = gtsam.LevenbergMarquardtOptimizer(graph, initial_estimate, parameters)

    # run optimizer
    result = optimizer.optimize()
